### DISCUSSION TITLE
assertions don't work except for single items?

### DISCUSSION THREAD

**Discussion by rchiodo** [microsoft/pyright#8312] [Q&A] [answered]
I'm adding typing to some classes in debugpy and hit a problem that you can see here:
https://pyright-play.net/?code=JYWwDg9gTgLgBDAFlApgQwCbAHYHMBQ%2BAxgDZoDO5cAKiuTABQCUAXPnB3APokREDWcALxwAchGwp2nHn36yBwhMnRY8AOgBKAGTnNpHLjnppsRFEvGSDcGxhQAzbl0kB3LlwalyAGjgAqNChcXwD-flcgkNYbTjgAYjgAQQA3CGAMOF4BHFw4CjgQAFciRHyqSEpgACMSC1dgJDhUMHQYFEyMCCLalABaUpQBKj7lCxAIeljORKIIEAnsODA0JuAqV0QUJZQUlCgATyRcuHX8klRMA-Ps1Y71aY5gJ2wIeG91I2wTMxQ2OIBMzgAEknEh9igAORUV7LfZ9GAHVpZOR%2BNAkW7tU4wB6AwHPOAfBSCM5WP6PPEcBpNInZeR0-6UplPJy0uSnKhkxnMnmEkjkT50pRIS5qXBaXQCfQUuKJcSuZSrOCuCyINB7fJwqAIpEWOl%2BFWE0xwcgHMzICTAABeFjm3xgUBKMGAElxPIJr3e-M%2BxhgpnM3N5ysaZTZAkDQc4HrefIFXx%2BAZlkdjPvt-osIgg1QAVkMcR43B4vPymEnI4lqBBlqg9th4GqUicPDhGh5Ttg0vwOoUiiRnWA6ghQHQ-ER0SQy0HEo04FtUHAsKgiDASAc0dhMuCli0yOZscGmmgTTAeoqYJPeYkunQ4J7ECdRkglfRTw0MXBcCh4GgijAIH0cE7bsJGaL8iigbAfAvHlElMTIQPHOAHF-cCLHIXI6j-JY0CIcxKDoN1kzxIlfXTT4vlbTxSyIykiQomA2xEMgQGqDAj0CYJQn8cJIk4lgxAkKQKVQE8IJTeM-V%2BQg4nsJwPG2dooCLcgUBIBwYkpAAibTJUEY4qHQvBMJAv9q12BTCQkIhwNQOt8lwugBW0zSKSFEREVaBgVLUphBTkCkKBU2AUUUM5PQEyRbEpOl1BwgBHIpgFQfRKRE8Clm8hxpM4WTnBQAAPSivNUhw-AKoguA8lAyvyiqUnRIpqrgcrKuqDS8WcgBVbAhX0k0MK-Eyq3HCAFTtayoFs78HModRnIC-Dgqq4qfL80KYRjMkorxZbMt84l1FQOoKBQZggA

Why does `__enter__` work, but `__exit__` has an error? Is this just a limitation of the assert type guarding?

**Comment by erictraut** [comment, marked as answer]
The expression `type(self)._lock is not None` is not a supported type guard pattern. For a full list of supported type guard patterns, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-guards).

In general, expressions that involve call expressions are not safe to narrow because calls can have side effects or return different values from one call to the next.

The code will type check without errors if you change it to the following:

```python
    def __exit__(self, exc_type, exc_value, exc_tb):
        """Unlock this singleton to allow concurrent access."""
        temp = type(self)
        assert temp._lock is not None 
        temp._lock.release()
```
